### PR TITLE
SaveGame::Save_459490()

### PR DESCRIPTION
**Pull request by d3nwah**
Includes/depends on https://github.com/AliveTeam/alive_reversing/pull/736

### CODE REVIEW

File: Source/AliveLibAO/SaveGame.cpp [hunk 1]
```diff
@@ -9,5 +15,232 @@ void SaveGame::sub_459970(BYTE*, int)
...
+
+    char path = 0;
+    if (gMap_507BA8.field_0_current_level == LevelIds::eRuptureFarmsReturn_13)
+    {
+        short path_id = -1;
+        for (int i = 0; i < 4; i++)
+        {
+            for(int j = 0; j < 8; j++ )
+            {
+                path = byte_4BC670[8 * i + j];
```
Comment (paulsapps) [inline Source/AliveLibAO/SaveGame.cpp:125]: this thing must be a 2d array

File: Source/AliveLibAO/SaveGame.cpp [hunk 2]
```diff
@@ -9,5 +15,232 @@ void SaveGame::sub_459970(BYTE*, int)
...
+    }
+    pSaveData->field_2AC = bUseAltSaveHeader_5076B4;
+    pSaveData->field_2AE = sCurrentControllerIndex_5076B8;
+    gMap_507BA8.SaveBlyData_446900(pSaveData->field_2B0_pSaveBuffer);
+    int* table = &pSaveData->field_204_zone_number;
+    int counter = 0;
+    for (int i = 1919; i > 0; i--)
+    {
+        table += 4;
+        counter += *table;
```
Comment (paulsapps) [inline Source/AliveLibAO/SaveGame.cpp:240]: its summing the 1919 fields after 200 and set it in 200, so seems like some pesudo crc/hash/anti save tampering thing

File: Source/AliveLibAE/QuikSave.cpp [hunk 3]
```diff
@@ -386,6 +386,7 @@ EXPORT void CCSTD Quicksave_SaveBlyData_4C9660(BYTE* pSaveBuffer)
             const int heightCount = (pPathData->field_6_bBottom - pPathData->field_2_bRight) / pPathData->field_C_grid_height;
             BYTE** ppPathRes = ResourceManager::GetLoadedResource_49C2A0(ResourceManager::Resource_Path, i, TRUE, FALSE);
             if (ppPathRes)
+            {
```
Comment (paulsapps) [inline Source/AliveLibAE/QuikSave.cpp:389]: shoryuken

Comment (d3nwah) [reply Source/AliveLibAE/QuikSave.cpp:389]: sure-I-can

File: Source/AliveLibAO/SaveGame.cpp [hunk 4]
```diff
@@ -1,13 +1,306 @@
...
+    {
+        pHeaderToUse = &sSaveHeader2_4CF2B0;
+    }
+
+    pSaveData->field_0_header = *pHeaderToUse;
+
+    //TODO NUKE CASTS
+    sub_45A2D0(
+        (BYTE *)&pSaveData->field_0_header.field_0_frame_1_name[4],
+        (BYTE *) &dword_4BC62C[static_cast<int>(gMap_507BA8.field_0_current_level)],
```
Comment (paulsapps) [inline Source/AliveLibAO/SaveGame.cpp:173]: redundant cast?

File: Source/AliveLibAO/SaveGame.cpp [hunk 5]
```diff
@@ -1,13 +1,306 @@
...
+    else
+    {
+        pHeaderToUse = &sSaveHeader2_4CF2B0;
+    }
+
+    pSaveData->field_0_header = *pHeaderToUse;
+
+    //TODO NUKE CASTS
+    sub_45A2D0(
+        (BYTE *)&pSaveData->field_0_header.field_0_frame_1_name[4],
```
Comment (paulsapps) [inline Source/AliveLibAO/SaveGame.cpp:172]: change type to avoid cast

File: Source/AliveLibAO/SaveGame.cpp [hunk 6]
```diff
@@ -1,13 +1,306 @@
...
+                {
+                    path_id = i;
+                    i = 4;
+                    break;
+                }
+            }
+        }
+        if (path_id != -1)
+        {
+            pSaveData->field_0_header.field_0_frame_1_name[44] = -127;
```
Comment (paulsapps) [inline Source/AliveLibAO/SaveGame.cpp:200]: I think these are unsigned, change to hex in IDA to figure out the values, prob 0x8X

File: Source/AliveLibAO/SaveGame.cpp [hunk 7]
```diff
@@ -1,13 +1,306 @@
...
+    }
+    pSaveData->field_22C_ah_health = sActiveHero_507678->field_100_health;
+    pSaveData->field_23C_ah_flipX = sActiveHero_507678->field_10_anim.field_4_flags.Get(AnimFlags::eBit5_FlipX);
+    pSaveData->field_230_ah_sprite_scale = sActiveHero_507678->field_BC_sprite_scale;
+    pSaveData->field_244_state = sActiveHero_507678->field_110_state;
+    pSaveData->field_248_gnFrame = sActiveHero_507678->field_114_gnFrame;
+    pSaveData->field_24C_field_118 = sActiveHero_507678->field_118;
+    pSaveData->field_250_throwable_count = sActiveHero_507678->field_19C_throwable_count;
+    pSaveData->field_253_scrabania_done = sActiveHero_507678->field_2A8_flags.Get(Flags_2A8::e2A8_eBit13_bScrabinaDone);
+    pSaveData->field_264 = -(gInfiniteGrenades_5076EC != 0);
```
Comment (paulsapps) [inline Source/AliveLibAO/SaveGame.cpp:246]: gInfiniteGrenades_5076EC  ? :-1 : 0 ? or whatever the right way around is

File: Source/AliveLibAO/Abe.cpp [hunk 8]
```diff
@@ -55,19 +55,7 @@ START_NS_AO;
...
-class Demo
-{
-public:
-    EXPORT static void CC Save_459490(BYTE* /*pBuffer*/)
-    {
-        // This is actually a game save
-        NOT_IMPLEMENTED();
-    }
-};
+ALIVE_ARY_EXTERN(SaveData, 8192, gSaveBuffer_505668);
```
Comment (paulsapps) [inline Source/AliveLibAO/Abe.cpp:58]: its not an array :O

File: Source/AliveLibAO/SaveGame.cpp [hunk 9]
```diff
@@ -1,13 +1,305 @@
...
+void SaveGame::sub_45A2D0(unsigned char*, const unsigned char* const* , int)
+{
+    NOT_IMPLEMENTED();
+}
+
+ALIVE_VAR_EXTERN(int, sGasTimer_507700);
+ALIVE_VAR_EXTERN(short, sRescuedMudokons_5076C0);
+ALIVE_VAR_EXTERN(short, sKilledMudokons_5076BC);
+
+ALIVE_ARY(1, 0x505668, SaveData, 1, gSaveBuffer_505668, {});
```
Comment (paulsapps) [inline Source/AliveLibAO/SaveGame.cpp:27]: should just be a struct not an array

**Pull request merged**